### PR TITLE
change flow to run as system test with local ER

### PR DESCRIPTION
**Pull request by evasiba-ericsson**
### Applicable Issues
Modify flow test to be able to run it locally with real Mongo, RabbitMq and ER.

### Description of the Change
Add a flag that controls where to run the test with real components or to mock them. Refactor linking between historical events used by the test.

### Alternate Designs
Use control flag as system property.

### Benefits
We can now test ER locally to see that upstream response will be compliant.

### Possible Drawbacks
None

### Sign-off

Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.

Signed-off-by: Vasile Baluta, vasile.baluta@ericsson.com

### CODE REVIEW

**Comment by e-edling-ericsson** [comment]
Right now we need to go into the test source code in order to change the systemTest variable right? Is this just temporary? I guess we will set it using command line arguments or some other kind of mechanism later?

File: src/test/java/com/ericsson/ei/flowtests/FlowTest.java [hunk 1]
```diff
@@ -65,20 +70,39 @@
...
+    		final URL upStreamInput = this.getClass().getClassLoader().getResource(UPSTREAM_INPUT_FILE);
+    		ArrayNode upstreamJson = (ArrayNode) objectMapper.readTree(upStreamInput);
+//    		rabbitMqTemplate.setQueue(null);
+    		if (upstreamJson != null) {
+    			for (JsonNode event : upstreamJson) {
+    				String eventStr = event.toString();
+    				rabbitMqTemplate.convertAndSend(eventStr);
+    			}    			
+    		}
+    		boolean wait = true;
```
Comment (a-breid-ericsson) [inline src/test/java/com/ericsson/ei/flowtests/FlowTest.java]: The boolean wait is never used.
There are commented code here that should be removed if not used.
There is a mix of tabs and spaces in this file.

File: src/test/java/com/ericsson/ei/flowtests/FlowTestBase.java [hunk 2]
```diff
@@ -77,25 +82,33 @@
...
 
     @Override
     public void beforeTestClass(TestContext testContext) throws Exception {
-        System.setProperty("flow.test", "true");
-        createFlowTestConfigs();
-        getFlowTestConfigs().init();
+    	if (!systemTest) {
+    		System.setProperty("flow.test", "true");
+    		createFlowTestConfigs();
+    		getFlowTestConfigs().init();
```
Comment (a-breid-ericsson) [inline src/test/java/com/ericsson/ei/flowtests/FlowTestBase.java]: There is a mix of spaces and tabs here

File: src/test/java/com/ericsson/ei/flowtests/FlowTestBase.java [hunk 3]
```diff
@@ -203,8 +227,14 @@ JsonNode getJSONFromFile(String filePath) throws IOException {
     }
 
     // count documents that were processed
-    private long countProcessedEvents(String database, String collection) {
-        MongoDatabase db = getFlowTestConfigs().getMongoClient().getDatabase(database);
+    private long countProcessedEvents(String database, String collection) {    	
+    	MongoClient mongoClient = null;
+    	if (systemTest) {
```
Comment (a-breid-ericsson) [inline src/test/java/com/ericsson/ei/flowtests/FlowTestBase.java]: Every where else there is if (!sy...) so the flow test code is at the above part of the if statement. Try to keep the same in the file so it is easier to follow the 2 different tests.

File: src/test/java/com/ericsson/ei/flowtests/FlowTest.java [hunk 4]
```diff
@@ -65,20 +70,39 @@
...
+    		JsonNode upstreamJson = objectMapper.readTree(upStreamResult);
+    		objectNode.set("upstreamLinkObjects", upstreamJson);
+    		objectNode.set("downstreamLinkObjects", objectMapper.createArrayNode());
+
+    		when(erQueryService.getEventStreamDataById(anyString(), any(SearchOption.class), anyInt(), anyInt(),
+    				anyBoolean())).thenReturn(new ResponseEntity<>(objectNode, HttpStatus.OK));
+    	} else {
+    		final URL upStreamInput = this.getClass().getClassLoader().getResource(UPSTREAM_INPUT_FILE);
+    		ArrayNode upstreamJson = (ArrayNode) objectMapper.readTree(upStreamInput);
+//    		rabbitMqTemplate.setQueue(null);
```
Comment (tobiasake) [inline src/test/java/com/ericsson/ei/flowtests/FlowTest.java]: If not used, then remove this line.

File: src/test/java/com/ericsson/ei/flowtests/FlowTestBase.java [hunk 5]
```diff
@@ -77,25 +82,33 @@
     private static ObjectMapper objectMapper = new ObjectMapper();
 
     private static HashMap<String, FlowTestConfigs> configsMap = new HashMap<String, FlowTestConfigs>();
+    
+    protected static boolean systemTest = false;
```
Comment (tobiasake) [inline src/test/java/com/ericsson/ei/flowtests/FlowTestBase.java]: Can't this be changed to property.
     protected static boolean systemTest = false; 
    @Value("${ei.backendInstancesFilePath:#{null}}")
    private Boolean systemTestProperty;

If (systemTestProperty) {
    systemTest = true;
} 

Then its default systemTest false and if systemTestProperty is provided, then its true.
Example: java -jar <appl name>.jar -systemTestProperty=true

Or something similar. Pick better property name, if you change to this solution.

Comment (evasiba-ericsson) [reply src/test/java/com/ericsson/ei/flowtests/FlowTestBase.java]: Done.

File: src/test/java/com/ericsson/ei/flowtests/FlowTest.java (hunk 4 above)
Comment (evasiba-ericsson) [reply src/test/java/com/ericsson/ei/flowtests/FlowTest.java]: Done.

File: src/test/java/com/ericsson/ei/flowtests/FlowTestBase.java (hunk 3 above)
Comment (evasiba-ericsson) [reply src/test/java/com/ericsson/ei/flowtests/FlowTestBase.java]: Done.

File: src/test/java/com/ericsson/ei/flowtests/FlowTestBase.java (hunk 2 above)
Comment (evasiba-ericsson) [reply src/test/java/com/ericsson/ei/flowtests/FlowTestBase.java]: Done.

File: src/test/java/com/ericsson/ei/flowtests/FlowTest.java (hunk 1 above)
Comment (evasiba-ericsson) [reply src/test/java/com/ericsson/ei/flowtests/FlowTest.java]: Done.

**Pull request merged**